# 4.5 Exercises

## Import libraries

In [5]:
# Import libraries
import pandas as pd
import numpy as np
import os

## Import Data

In [7]:
# Create Path to create folder into a string
path = r'C:\Users\E054746\OneDrive - RSM\Data Analytics\CareerFoundry\Data Immersion\Achievement 4\11-2021 Instacart Basket Analysis'

In [8]:
# Import Products Data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [9]:
# Import Orders Data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [10]:
# Create a test dataframe
df_test = pd.DataFrame()

In [11]:
# Create a mixed type column for test dataframe
df_test['mix'] = ['a', 'b', 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
# Find missing values in products dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Find the observations with missing values
df_nan = df_prods[df_prods['product_name'].isnull()== True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
# Count number of rows and columns
df_prods.shape

(49693, 5)

In [14]:
# Create new dataframe with non-missing values only
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
df_prods_clean.shape

(49677, 5)

In [15]:
# Find full duplicates in products dataframe 
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
df_prods_clean.shape

(49677, 5)

In [17]:
# Create new dataframe for products that doesn't include duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [18]:
df_prods_clean_no_dups.shape

(49672, 5)

# 4.5 Task Responses

### Question 2: Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [22]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### Response to question 2: For the columns product_id, aisle_id and department_id, the numbers should switch from floating-point numbers to integers. Also, under prices, I would find the product that has a price of 99,999 as this price seems excessive. 

### Question 3: Check for mixed-type data in your df_ords dataframe.

In [11]:
# Check for mixed type data in orders df
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### Question 4: If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

Response to Question 4: The command did not result in mix-type variables.

### Question 5: Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [13]:
# Find missing values in orders df
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [20]:
# Find the observations with missing values in orders df
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull()== True]

In [21]:
df_nan_ords

Unnamed: 0  order_id  user_id  order_number  order_dow  \
0                 0   2539329        1             1          2   
11               11   2168274        2             1          2   
26               26   1374495        3             1          1   
39               39   3343014        4             1          6   
45               45   2717275        5             1          3   
...             ...       ...      ...           ...        ...   
3420930     3420930    969311   206205             1          4   
3420934     3420934   3189322   206206             1          3   
3421002     3421002   2166133   206207             1          6   
3421019     3421019   2227043   206208             1          1   
3421069     3421069   3154581   206209             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

#### Response to Question 5: There are over 206,000 missing values under days_since_prior_order. The missing values here likely means that these customers are first-time customers purchasing from Instacart for the first time. 

### Question 6: Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

In [13]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_dow  order_hour_of_day  \
0           0   2539329        1             1          2                  8   
1           1   2398795        1             2          3                  7   
2           2    473747        1             3          3                 12   
3           3   2254736        1             4          4                  7   
4           4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

### Question 7: Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [22]:
# Find full duplicates in Orders DF
df_dups_ords = df_ords[df_ords.duplicated()]

In [17]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

Response to Question 7: No duplicates were found

### Question 8: Address the duplicates using an appropriate method. In a markdown cell, explain why you used your method of choice.

Response to Question 8: Because I did not find duplicates, no action will be necessary.

In [35]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [37]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

### Question 9: Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [38]:
 #Exporting orders data to prepared folder
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [39]:
 #Exporting products data to prepared folder
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))